### PR TITLE
Sigkill after grace period

### PR DESCRIPTION
**Pull request by BenElgar**

### CODE REVIEW

File: cmd/tick/main.go [hunk 1]
```diff
@@ -4,10 +4,24 @@ package main
...
+	"os/signal"
+	"syscall"
 	"time"
 )
 
 func main() {
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		for sig := range c {
```
Comment (brandonbloom) [inline cmd/tick/main.go:17]: I believe that will prevent the `tick` program from terminate on ^c in the terminal. This little command line tool is too useful to make it broken :)

If you were just doing this for testing one-off, I'd revert this change. If it's useful going forward, we should make another tool like "hang" or similar.

File: components/process/provider.go [hunk 2]
```diff
@@ -6,14 +6,16 @@ type Provider struct {
...
 type Spec struct {
-	Directory   string            `json:"directory"`
-	Program     string            `json:"program"`
-	Arguments   []string          `json:"arguments"`
-	Environment map[string]string `json:"environment"`
+	Directory                  string            `json:"directory"`
+	Program                    string            `json:"program"`
+	Arguments                  []string          `json:"arguments"`
+	Environment                map[string]string `json:"environment"`
+	ShutdownGracePeriodSeconds int               `json:"shutdownGracePeriodSeconds"`
```
Comment (brandonbloom) [inline components/process/provider.go:13]: This is a mouthful. Why not call it `shutdownTimeout`?

Also: Why seconds and not milliseconds? Seems like 200 ms would be more than enough grace period for many use cases.

File: supervise/main.go [hunk 3]
```diff
@@ -78,31 +82,35 @@ MSGID = The message "type". Set to "out" or "err" to specify which stdio
...
+				if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
 					break
 				}
+				// After some timeout, check if the process is still running and, if so,
+				// send a SIGKILL.
+				time.Sleep(time.Second * time.Duration(timeout))
+				cmd.Process.Kill()
 			// Exit when child exits.
 			case syscall.SIGCHLD:
 				os.Exit(1)
```
Comment (brandonbloom) [inline supervise/main.go:109]: is this now potentially the wrong status code for the supervisor?

File: supervise/main.go [hunk 4]
```diff
@@ -78,31 +82,35 @@ MSGID = The message "type". Set to "out" or "err" to specify which stdio
 		fatalf("reporting pid: %v", err)
 	}
 
+	// Dial syslog.
+	conn, err := net.DialUDP("udp", nil, udpAddr)
```
Comment (brandonbloom) [inline supervise/main.go:86]: curious why this moved before the signal handling. I _think_ DialUDP will never block, but my intuition is to always configure signal handlers as early as possible in case blocking does happen.

File: supervise/main.go [hunk 5]
```diff
@@ -24,7 +24,7 @@ var varDir string
 
 func Main(command string, args []string) {
 	if len(args) < 4 {
-		fatalf(`usage: %s <address> <component-id> <working-directory> <program> <args...>
+		fatalf(`usage: %s <address> <component-id> <working-directory> <timeout> <program> <args...>
```
Comment (brandonbloom) [inline supervise/main.go:27]: Yeah, we should definitely do #47

File: supervise/main.go [hunk 6]
```diff
@@ -45,8 +45,12 @@ MSGID = The message "type". Set to "out" or "err" to specify which stdio
 	address := args[0]
 	componentID := args[1]
 	wd := args[2]
-	program := args[3]
-	arguments := args[4:]
+	timeout, timeoutErr := strconv.Atoi(args[3])
+	if timeoutErr != nil {
+		panic(timeoutErr)
```
Comment (brandonbloom) [inline supervise/main.go:50]: Use fatalf so that the error can be more easily reported back to the code that started the supervisor.

File: components/process/process.go [hunk 7]
```diff
@@ -41,7 +41,7 @@ func (provider *Provider) Start(ctx context.Context, input *core.StartInput) (*c
 }
 
 func (provider *Provider) start(ctx context.Context, componentID string, inputSpec string) (State, error) {
-	var spec Spec
+	spec := Spec{ShutdownGracePeriodSeconds: 5}
```
Comment (brandonbloom) [inline components/process/process.go:44]: This seems really subtle! I was confused at first how this was even working (assuming it is).

Seems like this will default the field when `UnmarshalString` runs only by virtue of the fact that if the field is missing, then it won't get overridden! It's not obvious to me that it wouldn't potentially be overwritten to 0.

A much clearer thing to do would be to change the field to a pointer type and then explicitly handle defaulting when the timeout is nil.

File: supervise/main.go [hunk 8]
```diff
@@ -78,31 +82,35 @@ MSGID = The message "type". Set to "out" or "err" to specify which stdio
...
 		for sig := range c {
 			switch sig {
 			// Forward signals to child.
-			case os.Interrupt, os.Kill:
-				if err := cmd.Process.Signal(sig); err != nil {
+			case os.Interrupt, syscall.SIGTERM:
+				if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
 					break
 				}
+				// After some timeout, check if the process is still running and, if so,
```
Comment (brandonbloom) [inline supervise/main.go:103]: This comment suggests some checks/conditions, but it seems to um.. be unconditionally :-P Tho I think unconditional is fine here, assuming no immediate pid reuse (and I think there is nothing better you can do beyond that).

File: components/process/process.go [hunk 9]
```diff
@@ -72,6 +72,7 @@ func (provider *Provider) start(ctx context.Context, componentID string, inputSp
 			provider.SyslogAddr,
 			componentID,
 			provider.WorkspaceDir,
+			fmt.Sprintf("%d", spec.ShutdownGracePeriodSeconds),
```
[the remark this answers is not in the document]
Comment (BenElgar) [reply components/process/process.go:75]: No this is a conversion to a string

File: cmd/tick/main.go (hunk 1 above)
Comment (BenElgar) [reply cmd/tick/main.go:17]: Fair enough

File: components/process/provider.go (hunk 2 above)
Comment (BenElgar) [reply components/process/provider.go:13]: I basically took the name from Kubernetes which calls it the same thing [0]. `shutdownTimeout` sounds a bit like the maximum amount of time that the process should be allowed to run for.

There aren't any use cases which I can think of where 200ms would be better than 0 or 1 second. If a process intercepts the signal it is presumably going to do some I/O which I would always have more than a 1 second timeout for. For context, the default Kubernetes grace period is 30 seconds.

[0] https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution

File: supervise/main.go (hunk 8 above)
Comment (BenElgar) [reply supervise/main.go:85]: Fixed — I removed the check when I was debugging and forgot to add it back.

File: components/process/process.go (hunk 7 above)
Comment (BenElgar) [reply components/process/process.go:44]: Discussed in person but it seems reasonably common. It's first answer in both of the top two results on Google for "go json unmarshal default".
https://eli.thegreenplace.net/2020/optional-json-fields-in-go/#setting-default-values
https://stackoverflow.com/a/30445480

File: supervise/main.go (hunk 6 above)
Comment (BenElgar) [reply supervise/main.go:50]: :+1:

File: supervise/main.go (hunk 4 above)
Comment (BenElgar) [reply supervise/main.go]: Ah this is a leftover from when I was using the connection for printing debug messages. I moved the signal handling above even starting the process. The case in which the signal occurs before the process has been started seems to be less important than the case when the process has been started but signal forwarding hasn't been set up.

File: components/process/process.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (brandonbloom) [reply components/process/process.go:75]: haha whoops misread :)

File: supervise/main.go (hunk 3 above)
Comment (BenElgar) [reply supervise/main.go:96]: Fixed
